### ISSUE TITLE
Invoke-WebRequest: specifying Content-Length gives duplicate values and raises false error regarding -ContentType

### ISSUE THREAD

**Issue by jorins** [PowerShell/PowerShell#12500]
# Steps to reproduce

## 1. Not specifying Content-Length

This gives us expected success:

```powershell
Invoke-WebRequest -Method PUT -URI "http://127.0.0.1:8080/put" -Body ':)'
```

gives

```none
PS C:\> ncat -lp 8080
PUT /put HTTP/1.1
Host: 127.0.0.1:8080
User-Agent: Mozilla/5.0 (Windows NT 10.0; Microsoft Windows 10.0.18363; en-SE) PowerShell/7.0.0
Content-Length: 2

:)
```

## 2. Specifying Content-Length

```powershell
Invoke-WebRequest -Method PUT -URI "http://127.0.0.1:8080/put" -Headers @{"Content-Length" = 2} -Body ':('
```
gives

> Invoke-WebRequest: The cmdlet cannot run because the -ContentType parameter is not a valid Content-Type header. Specify a valid Content-Type for -ContentType, then retry. To suppress header validation, supply the -SkipHeaderValidation parameter.

## 3. Adding -ContentType to address the supposed error

```powershell
Invoke-WebRequest -Method PUT -URI "http://127.0.0.1:8080/put" -Headers @{"Content-Length" = 2} -Body ':(' -ContentType "text/plain"
```
gives us the same

> Invoke-WebRequest: The cmdlet cannot run because the -ContentType parameter is not a valid Content-Type header. Specify a valid Content-Type for -ContentType, then retry. To suppress header validation, supply the -SkipHeaderValidation parameter.

## 4. Skipping header validation to see what happens:
```powershell
Invoke-WebRequest -Method PUT -URI "http://127.0.0.1:8080/put" -Headers @{"Content-Length" = 2} -Body ':(' -SkipHeaderValidation
```
reveals that the Content-Length header is not overridden, but rather added to as if it was a list:
```none
PS C:\> ncat -lp 8080
PUT /put HTTP/1.1
Host: 127.0.0.1:8080
User-Agent: Mozilla/5.0 (Windows NT 10.0; Microsoft Windows 10.0.18363; en-SE) PowerShell/7.0.0
Content-Length: 2, 2
:(
```

# Expected behavior
**Testing with PowerShell 5 for reference of expected behaviour**

## 1. Content-Length is correct

```powershell
Invoke-WebRequest -Method PUT -URI "http://127.0.0.1:8080/put" -Headers @{"Content-Length" = "2"} -ContentType "application/json" -Body ":("
```
gives
```none
PS C:\> ncat -lp 8080
PUT /put HTTP/1.1
User-Agent: Mozilla/5.0 (Windows NT; Windows NT 10.0; en-US) WindowsPowerShell/5.1.18362.752
Content-Type: application/json
Host: 127.0.0.1:8080
Content-Length: 2
Connection: Keep-Alive

:(
```

## 2. Content-Length too high
```powershell
Invoke-WebRequest -Method PUT -URI "http://127.0.0.1:8080/put" -Headers @{"Content-Length" = "5"} -ContentType "application/json" -Body ":("
```
raises 

```
Invoke-WebRequest : You must write ContentLength bytes to the request stream before calling [Begin]GetResponse.
At line:1 char:1
+ Invoke-WebRequest -Method PUT -URI "http://127.0.0.1:8080/put" -Heade ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Invoke-WebRequest], ProtocolViolationException
    + FullyQualifiedErrorId : System.Net.ProtocolViolationException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
```
and gives
```
PS C:\> ncat -lp 8080
PUT /put HTTP/1.1
User-Agent: Mozilla/5.0 (Windows NT; Windows NT 10.0; en-US) WindowsPowerShell/5.1.18362.752
Content-Type: application/json
Host: 127.0.0.1:8080
Content-Length: 5
Connection: Keep-Alive

:(
```

## 3. Content-Length too low

```powershell
Invoke-WebRequest -Method PUT -URI "http://127.0.0.1:8080/put" -Headers @{"Content-Length" = "1"} -ContentType "application/json" -Body ":("
```
raises
```
Invoke-WebRequest : Bytes to be written to the stream exceed the Content-Length bytes size specified.
At line:1 char:1
+ Invoke-WebRequest -Method PUT -URI "http://127.0.0.1:8080/put" -Heade ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Invoke-WebRequest], ProtocolViolationException
    + FullyQualifiedErrorId : System.Net.ProtocolViolationException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
```
and gives
```
PS C:\> ncat -lp 8080
PUT /put HTTP/1.1
User-Agent: Mozilla/5.0 (Windows NT; Windows NT 10.0; en-US) WindowsPowerShell/5.1.18362.752
Content-Type: application/json
Host: 127.0.0.1:8080
Content-Length: 1
Connection: Keep-Alive

```

# Actual behavior

cmdlet does not run:
> Invoke-WebRequest: The cmdlet cannot run because the -ContentType parameter is not a valid Content-Type header. Specify a valid Content-Type for -ContentType, then retry. To suppress header validation, supply the -SkipHeaderValidation parameter.

If `-SkipHeaderValidation` is passed, we can see that the `Content-Length` header appends the value specified to the calculated value like a list, as opposed to overriding it:

```
PS C:\> ncat -lp 8080
PUT /put HTTP/1.1
Host: 127.0.0.1:8080
User-Agent: Mozilla/5.0 (Windows NT 10.0; Microsoft Windows 10.0.18363; en-SE) PowerShell/7.0.0
Content-Length: 2, 2
:(
```

# Environment data
```
Name                           Value
----                           -----
PSVersion                      7.0.0
PSEdition                      Core
GitCommitId                    7.0.0
OS                             Microsoft Windows 10.0.18363
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by iSazonov** [comment]
.Net Core HttpClient makes all validations and PowerShell follows the .Net behavior.

**Comment by vexx32** [comment]
So either this is a bug in how we're using the API, or it's a bug in .NET. If it's a bug in .NET we should look to find the simplest reproducible case with only .NET code and move the issue to the dotnet/runtime repo.

**Comment by iSazonov** [comment]
I don't understand the scenario to force setting Content-Length header.

**Comment by jorins** [comment]
The issue came to light when working with Azure blob storage when we had to generate an authorization header for a PUT request. That requires us to include content length, so we opted to pass it via the headers object that we also use in the request. We were able to work around it by passing content length in a separate parameter to the authorization header generation function instead.

It does of course raise a few questions as to what the ideal approach here is. On one hand, forcefully setting the content length header is probably more prone to errors. On the other hand, since we have to pass it manually to the authorization token generation function, putting it in the headers as well will make sure that there is no mismatch between the authorization header and the content length header.

Regardless, this bug is what it is. The workaround is trivial in our case but the cause took me a good while to isolate. Hope that explains our scenario!
-------- Original Message --------

**Comment by iSazonov** [comment]
.Net API add the Content-Length automatically. I don't understand why you need to calculate the length in your code. If .Net API do wrong the length it is a bug there.

**Comment by jorins** [comment]
Before making the request, we have to sign an authorization header, which requires we know the content length. For example, to make a PUT request, we use roughly this code:

```powershell
function Set-AzBlob
{
    param(
        <... snip ...>
        [Parameter(Mandatory=$true)][string]$BlobName
        [Parameter(Mandatory=$true)][string]$Contents
    )

    <... snip ...>

    $Headers = GetCommonHeaders
    $Headers["Content-Length"] = GetContentLength $Contents
    $Headers["Authorization"] = <... snip ...> PUT $Headers $ContentLength

    Invoke-WebRequest -Method PUT -URI $URI -Headers $Headers
}

function GetAuthorizationHeader
{
    param (
        <... snip ...>
        $method,
        $headers,
    )

    <... snip ...>

    $StringToSign = $method.ToUpper()   + "`n" +
               "`n" +
               "`n" +
               $headers["Content-Length"] + "`n" +
               "`n" +
               $headers["Content-Type"]   + "`n" +
               "`n" +
               "`n" +
               $headers["If-Match"]      + "`n" +
               $headers["If-None-Match"] + "`n" +
               "`n" +
               "`n" +
               $CanonicalizedHeaders + "`n" +
               $CanonicalizedResource

    return sign($StringToSign)
}

```

This wasn't working, and I traced the issue to the incorrect `Content-Length` header. The workaround was to pass `Content-Length` separately to the GetAuthorizationHeader function:

```powershell
function Set-AzBlob
{
    param(
        <... snip ...>
        [Parameter(Mandatory=$true)][string]$BlobName
        [Parameter(Mandatory=$true)][string]$Contents
    )

    <... snip ...>

    $Headers = GetCommonHeaders
    $ContentLength = GetContentLength $Contents
    $Headers["Authorization"] = <... snip ...> PUT $Headers $ContentLength

    Invoke-WebRequest -Method PUT -URI $URI -Headers $Headers
}

function GetAuthorizationHeader
{
    param (
        <... snip ...>
        $method,
        $headers,
        $ContentLength
    )

    <... snip ...>

    $StringToSign = $method.ToUpper()   + "`n" +
               "`n" +
               "`n" +
               $ContentLength + "`n" +
               "`n" +
               $headers["Content-Type"]   + "`n" +
               "`n" +
               "`n" +
               $headers["If-Match"]      + "`n" +
               $headers["If-None-Match"] + "`n" +
               "`n" +
               "`n" +
               $CanonicalizedHeaders + "`n" +
               $CanonicalizedResource

    return sign($StringToSign)
}
```

**Comment by iSazonov** [comment]
@jorins Thanks for clarify! I think you could ask .Net team in .Net Runtime repository about HttpClient behavior. I guess it is by-design but you could discuss this with .Net team.

**Comment by iSazonov** [comment]
Close as stale issue. Feel free to discuss this with .Net team.

**Comment by thezim** [comment]
While it is understood that Content-Length should not be added manually the error message should still be  addressed. If anyone accidentally adds any custom headers multiple times by accident times they will get the cryptic `Content-Type` error which is inaccurate.

I spent way too much time troubleshooting this failure when trying to consume Google services. Per the error message I told it to skip the `Content-Type` check which then just yielded 'BadRequest' from Google. I thought I had malformed my request body until I ran the same code on Windows PowerShell as it worked, even with my manual `Content-Length` header added.

The offending code is [here](https://github.com/PowerShell/PowerShell/blob/c3fa70380c465c21023fa9fb0842e7a5a31cfb17/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs#L1282). I will try to get a PR up when I have some free time.

This issue should be reopened. @iSazonov
